File: backend/app/services/lowest_price.py
from app.models.common import SourcePlatform
from app.models.property import PriceObservation


class LowestPriceEngine:
    def attach_lowest_price(self, property_doc: dict) -> dict:
        observations = [
            PriceObservation(source=property_doc["source_platform"], rent=property_doc["rent"], url=property_doc.get("source_url"))
        ]
        observations.extend(PriceObservation.model_validate(item) for item in property_doc.get("price_history", []))
        lowest = min(observations, key=lambda item: item.rent)
        property_doc["lowest_price"] = lowest.model_dump()
        property_doc["source_platform"] = SourcePlatform(property_doc["source_platform"])
        return property_doc

File: backend/src/services/lowestPrice.ts
import { PriceObservation } from "../models/property.js";
import { attachProviderUrls, resolveProviderListingUrl } from "./providerUrls.js";

export class LowestPriceEngine {
  attachLowestPrice(propertyDoc: Record<string, any>): Record<string, any> {
    attachProviderUrls(propertyDoc);
    const observations: PriceObservation[] = [
      {
        source: propertyDoc.source_platform,
        rent: propertyDoc.rent,
        url: propertyDoc.listing_url || propertyDoc.source_url || propertyDoc.provider_url || null,
        observed_at: propertyDoc.created_at || new Date().toISOString(),
      },
    ];

    if (Array.isArray(propertyDoc.price_history)) {
      for (const item of propertyDoc.price_history) {
        if (item && typeof item.rent === "number") {
          observations.push({
            source: item.source,
            rent: item.rent,
            url:
              item.url ||
              resolveProviderListingUrl({
                source: item.source,
                city: propertyDoc.city,
                locality: propertyDoc.locality,
                title: propertyDoc.title,
              }) ||
              null,
            observed_at: item.observed_at || new Date().toISOString(),
          });
        }
      }
    }

    let lowest = observations[0];
    for (const obs of observations) {
      if (obs.rent < lowest.rent) lowest = obs;
    }

    propertyDoc.lowest_price = lowest;
    return propertyDoc;
  }
}
